_posts/2020-12-18-react-generic-context.md

```diff
@@ -0,0 +1,232 @@
+Mostly because the `AuthConsumer` and the `AuthProvider`  exposed by the core library had to point to a specific context object, so a function creating this context didn’t make sense.
+My knowledge of React is still limited and I decided to call it a day and submit the merge request for review.
+
+As I feared, type safety was a concern and had to be taken care of.
+Luckily we have a team with members with a lot more React experience and they guided me to a solution.
+
+## The Solution
+<img class="image fit center" alt="A picture of the holy grail" src="/img/react-generic-context/the-grail.jpg">
+
+When looking back, the solution is fairly simple, but elegant.
+Instead of letting the core library expose actual components, we let it expose a factory function.
+This factory function would create a context and in its turn call factory functions of each component and pass the context to it.
+Finally the function would just return the created components.
+
+```typescript
+// authentication-core/context/AuthContext.tsx
+import React from 'react';
+import { AuthContextProps } from '../domain/auth-context';
+
+interface AuthContextProps<T = unknown> {
+  signIn: (args?: unknown) => Promise<void>;
+  signOut: (args?: unknown) => Promise<void>;
+  userData?: T | null;
+  isExpired?: boolean;
+  status: 'loading' | 'idle';
+}
+
+export function createAuthContext<T = unknown>() {
+  return React.createContext<AuthContextProps<T> | null>(null);
+}
+```
+
+```typescript
+// authentication-core/factory.ts
+import { createAuthConsumer } from './components/AuthConsumer';
+import { createProtectedRoute } from './components/ProtectedRoute';
+import { createWithAuth } from './components/withAuth';
+import { createAuthContext } from './context/AuthContext';
+import { createUseAuth } from './hooks/useAuth';
+import { createAuthProvider } from './providers/AuthProvider';
+import { AuthService } from './services/auth.service';
+
+export function createAuthService<T = unknown>(authService: AuthService<T>) {
+  const AuthContext = createAuthContext<T>(); // We type the Context based on the generic AuthService
+  const useAuth = createUseAuth(AuthContext);
+  const AuthProvider = createAuthProvider(AuthContext, authService);
+  const AuthConsumer = createAuthConsumer(AuthContext);
+  return {
+    AuthContext,
+    AuthProvider,
+    AuthConsumer,
+    useAuth,
+  };
+}
+```
+
+The implementation libraries had to be adapted as well.
+They too expose a factory function, calling the core factory function with some pre-setup data and their respective service.
+
+```typescript
+// authentication-oidc/factory.ts
+import { createAuthService } from '@our-scope/authentication-core';
+import { OidcProps } from './domain/oidc-props';
+import { OidcAuthService } from './services/oidc-auth.service';
+
+export function createOidcAuthService(oidcProps: OidcProps) {
+  const oidcService = new OidcAuthService(oidcProps);
+  return createAuthService(oidcService);
```

Comment:
Should this createAuthService get a type now? Or is that inferred somehow from the oidcService (if that's the case, it's not superclear)?

Reply:
It's inferred. I'll mention it